src/Symfony/Component/HttpClient/RetryableHttpClient.php

```
@@ -54,40 +55,72 @@ public function __construct(HttpClientInterface $client, RetryDeciderInterface $
+        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options, &$retryCount, &$content, &$firstChunk) {
             $exception = null;
             try {
                 if ($chunk->isTimeout() || null !== $chunk->getInformationalStatus()) {
                     yield $chunk;
 
                     return;
                 }
-
-                // only retry first chunk
-                if (!$chunk->isFirst()) {
-                    $context->passthru();
-                    yield $chunk;
-
-                    return;
-                }
             } catch (TransportExceptionInterface $exception) {
                 // catch TransportExceptionInterface to send it to strategy.
             }
 
-            $statusCode = $context->getStatusCode();
-            $headers = $context->getHeaders();
-            if ($retryCount >= $this->maxRetries || !$this->decider->shouldRetry($method, $url, $options, $partialResponse = new MockResponse($context->getContent(), ['http_code' => $statusCode, 'headers' => $headers]), $exception)) {
+            if ($retryCount >= $this->maxRetries) {
                 $context->passthru();
                 yield $chunk;
 
                 return;
             }
 
+            $statusCode = $context->getStatusCode();
+            $headers = $context->getHeaders();
+            if ($chunk instanceof FirstChunk || null !== $exception) {
+                $shouldRetry = $this->decider->shouldRetry($method, $url, $options, $context->getStatusCode(), $context->getHeaders(), null, $exception);
+
+                if (false === $shouldRetry) {
+                    $context->passthru();
+                    yield $chunk;
+
+                    return;
+                }
+
+                // Decider need body to decide
+                if (null === $shouldRetry) {
```

When `null !== $exception`, we should throw a logic exception because there's no point waiting for a body then.

---

good catch!